include/envoy/upstream/upstream.h

```diff
@@ -168,6 +171,70 @@ class HostSet {
    * @return same as hostsPerLocality but only contains healthy hosts.
    */
   virtual const std::vector<std::vector<HostSharedPtr>>& healthyHostsPerLocality() const PURE;
+
+  /**
+   * Updates the hosts in a given host set.
+   *
+   * @param hosts supplies the (usually new) list of hosts in the host set.
+   * @param healthy hosts supplies the subset of hosts which are healthy.
+   * @param hosts_per_locality supplies the hosts subdivided by locality.
```

Comment:
At a high level, without diving into implementation, it seems that it's redundant to track and supply both (healthy) hosts and (healthy) hosts per locality, one of these (presumably the latter) should be sufficient. I know this is an extant thing, just wanted to start a thread to discuss this with you and @mattklein123 .

Reply:
Agree, but as it turns out we track locality inconsistently (I think the rule is only if there are multiple localities and one is local and it's configured?) so it's quite non-trivial to combine the two.

I wonder if the better move might be to always just have the "locality style" group and if we don't care to pick by locality it's just in one big group.  I know the healthy host is used for some things (mainly size for determining if it's a global panic, but size could be latched) but I can't remember if the full host list is ever used when splitting up by locality.

Reply:
Previously this was hidden inside the implementation and is split out like this for perf reasons (basically compute once and then use everywhere). @alyssawilk do we need to expose this at the interface layer? I agree with @htuch at the interface level it's not great.

Reply:
It's that, or doing a bunch of casts, since the generic priority set impl is handing out containers of HostSets not HostSet Impls.  And because they're complex types (it's super handy to iterate over the vector) I believe we can't just override.

If we prefer casts in upstream_impl.cc I'm happy to do that.

Reply:
I haven't thought through all the details but if you think this is the way to go it's fine with me. I  can take a more complete look through this tomorrow if that would help. Sorry for the delay.